# Data Consistency Checks

## Contents: Checking data for inconsistencies

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [3]:
# importing orders df

df_orders = pd.read_csv(r'C:\Users\thoms\Desktop\Data Analytics\Achievement 4\Instacart Basket Analysis\02 Data\Original Data\orders.csv', index_col = False )

In [4]:
# importing products

df_prods = pd.read_csv (r'C:\Users\thoms\Desktop\Data Analytics\Achievement 4\Instacart Basket Analysis\02 Data\Original Data\products.csv' , index_col = False)

In [5]:
path = r'C:\Users\thoms\Desktop\Data Analytics\Achievement 4\Instacart Basket Analysis\02 Data'

In [8]:
# Check dfs 

df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [9]:
df_orders.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [10]:
# Creat test df

df_test = pd.DataFrame ()

In [11]:
df_test['mix'] = ['a', 'b', 1, True ]

In [12]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [13]:
# finding the type

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [14]:
#change value to string

df_test['mix'] = df_test['mix'].astype('str')


In [15]:
#finding missing values

df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [16]:
df_nan = df_prods [df_prods ['product_name'] .isnull() == True]

In [17]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [18]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [19]:
df_prods.shape

(49693, 5)

In [20]:
# finding duplicates

df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [21]:
df_prods_clean.shape

(49677, 5)

In [22]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [23]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [24]:
# addressing duplicates

df_prods.drop_duplicates

<bound method DataFrame.drop_duplicates of        product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1         

In [25]:
df_prods_clean.shape

(49677, 5)

In [26]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()


In [27]:
df_prods_clean_no_dups.shape

(49672, 5)

In [28]:
df_prods_clean_no_dups.to_csv(os.path.join(path,'Prepared Data', 'products_checked.csv'))

# Task 2.9

## Question 2: Run the df.describe() function on your df_prods dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.


In [29]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [30]:
df_prods['prices']

0         5.8
1         9.3
2         4.5
3        10.5
4         4.3
         ... 
49688     5.3
49689     3.1
49690     7.8
49691     4.7
49692    13.5
Name: prices, Length: 49693, dtype: float64

In [31]:
df_prods.loc[(df_prods['prices'] == 99999)]

product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0

In [32]:
df_prods = df_prods.drop(labels=33666, axis=0)

In [33]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49692.000000  49692.000000   49692.000000  49692.000000
mean   24844.167653     67.769923      11.728347      7.981961
std    14343.807161     38.317091       5.850309     66.938457
min        1.000000      1.000000       1.000000      1.000000
25%    12422.750000     35.000000       7.000000      4.100000
50%    24844.500000     69.000000      13.000000      7.100000
75%    37265.250000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  14900.000000

### Answer: The values that are in the column Prices is extremely high for a grocery store. There needs to be an investigation into this column. 

 ## Question 3: Check for mixed-type data in your df_ords dataframe.

In [36]:
for col in df_orders.columns.tolist():
  weird = (df_orders[[col]].applymap(type) != df_orders[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_orders[weird]) > 0:
    print (col)

### Answer: Nothing is printed, therefore there are no mixed values in the df_orders dataframe.

## Question 4: If you find mixed-type data, fix it. The column in question should contain observations of a single data type

### Answer: Nothing is printed, therefore there are no mixed values in the df_orders dataframe.

## Question: 5 Run a check for missing values in your df_ords dataframe.

In [38]:
df_orders.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [39]:
df_nan = df_orders [df_orders ['days_since_prior_order'] .isnull() == True]

In [40]:
df_nan

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
11        2168274        2    prior             1          2   
26        1374495        3    prior             1          1   
39        3343014        4    prior             1          6   
45        2717275        5    prior             1          3   
...           ...      ...      ...           ...        ...   
3420930    969311   206205    prior             1          4   
3420934   3189322   206206    prior             1          3   
3421002   2166133   206207    prior             1          6   
3421019   2227043   206208    prior             1          1   
3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

### Answer: There is only one column in the df_orders dataframe that has missing values and this is the 'days_since_prior_order'. This can because this is the first and only order of the 206209 customers. Therefore, they will not have a prior order date ti give this column a value. 

## Question 6: Address the missing values using an appropriate method.

### Answer: No action should be taken on these rows with missing information, as this is important information for the Dataframe. What we could do is add a row for identifying first time orders from customers. Thus, creating a new varible to act as a flag based on the missing value. 

In [42]:
df_orders_clean = df_orders

In [43]:
df_orders_clean['initial_order'] = df_orders['days_since_prior_order'].isnull() == True

In [44]:
df_orders_clean

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
1         2398795        1    prior             2          3   
2          473747        1    prior             3          3   
3         2254736        1    prior             4          4   
4          431534        1    prior             5          4   
...           ...      ...      ...           ...        ...   
3421078   2266710   206209    prior            10          5   
3421079   1854736   206209    prior            11          4   
3421080    626363   206209    prior            12          1   
3421081   2977660   206209    prior            13          1   
3421082    272231   206209    train            14          6   

         order_hour_of_day  days_since_prior_order  initial_order  
0                        8                     NaN           True  
1                        7                    15.0          False  
2                       12                    21.0          False  
3                        7                    29.0          False  
4                       15                    28.0          False  
...                    ...                     ...            ...  
3421078                 18                    29.0          False  
3421079                 10                    30.0          False  
3421080                 12                    18.0          False  
3421081                 12                     7.0          False  
3421082                 14                    30.0          False  

[3421083 rows x 8 columns]

In [45]:
df_orders_clean.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  initial_order  
0                     NaN           True  
1                    15.0          False  
2                    21.0          False  
3                    29.0          False  
4                    28.0          False

## Question 7: Run a check for duplicate values in your df_ords data.

In [47]:
df_dups = df_orders_clean[df_orders_clean.duplicated()]

In [48]:
df_dups

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order, initial_order]
Index: []

### Answer: There are no duplicates in the df_orders dataframe. 

## Question 8: Address the duplicates using an appropriate method

In [49]:
df_dups.shape

(0, 8)

### Answer: There are no duplicates in the df_orders dataframe. Therefore, no action can or needs to be taken against the dataframe. 

## Question 9: Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [50]:
# df_prods

df_prods_clean.to_csv(os.path.join(path,'Prepared Data', 'prods_checked.csv'))

In [51]:
# df_orders

df_orders_clean.to_csv(os.path.join(path,'Prepared Data', 'ords_wrangled.csv'))